### PR TITLE
average completion time for action items

### PR DESCRIPTION
**Pull request by chance-cyphers**

### CODE REVIEW

File: migrations/1593194295169_add-status-to-action-items.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable @typescript-eslint/camelcase */
+import { MigrationBuilder, ColumnDefinitions } from 'node-pg-migrate';
+
+export const shorthands: ColumnDefinitions | undefined = undefined;
+
+export async function up(pgm: MigrationBuilder): Promise<void> {
+    pgm.addColumn('action_items', {
+        status: { type: 'varchar(20)' }
```
Comment (tchupp) [inline migrations/1593194295169_add-status-to-action-items.ts:8]: We should just use `text` as the column type. The only reason to use `varchar(X)` is if we want to force the text to be smaller than the given size

Comment (chance-cyphers) [reply migrations/1593194295169_add-status-to-action-items.ts:8]: is it not more performant or anything? Either way, sounds fine to me

File: migrations/1593197631571_add-closed-time-to-action-items.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable @typescript-eslint/camelcase */
+import { MigrationBuilder, ColumnDefinitions } from 'node-pg-migrate';
+
+export const shorthands: ColumnDefinitions | undefined = undefined;
+
+export async function up(pgm: MigrationBuilder): Promise<void> {
+    pgm.addColumn('action_items', {
+        closed_at: { type: 'timestamp' }
```
Comment (tchupp) [inline migrations/1593197631571_add-closed-time-to-action-items.ts:8]: do we want to make a distinction between "completed" and "canceled"?

File: migrations/1593194295169_add-status-to-action-items.ts (hunk 1 above)
Comment (chance-cyphers) [reply migrations/1593194295169_add-status-to-action-items.ts:8]: google tells me it mainly only matters if you want to put an index on the column. Which we don't. TIL

File: migrations/1593197631571_add-closed-time-to-action-items.ts (hunk 2 above)
Comment (chance-cyphers) [reply migrations/1593197631571_add-closed-time-to-action-items.ts:8]: yeah. I was going to have status be OPEN | COMPLETED | CANCELED, where the last two are types of "closed". So when there's a `closed_at` timestamp, there will also be a competed or canceled status

Comment (chance-cyphers) [reply migrations/1593197631571_add-closed-time-to-action-items.ts:8]: also lazily not representing that in another table or anything 🙃

File: src/handlers/event.handler.ts [hunk 3]
```diff
@@ -23,20 +24,31 @@ export const eventHandler = (client: Client) => {
...
-    const blocks = res.rows.map(row => markdownSection(row.description));
+    const completeds = result.rows.filter(row => row.status === 'COMPLETED');
 
-    const homeViewBlocks = homeView(blocks);
+    const avgMs = completeds.length > 0 ?
+        completeds
+            .map(row => row.closed_at.getTime() - row.created_at.getTime())
+            .reduce((acc, current) => acc + current) / completeds.length
+        : 0;
+    const averageTimeString = moment.duration(avgMs).humanize();
```
Comment (tchupp) [inline src/handlers/event.handler.ts:45]: This _might_ be slightly easier to read:
```
const avgMs = completeds.length <= 0 
    ? 0
    : completeds
        .map(row => row.closed_at.getTime() - row.created_at.getTime())
        .reduce((acc, current) => acc + current) / completeds.length
    ;
const averageTimeString = moment.duration(avgMs).humanize();
```

File: src/view.ts [hunk 4]
```diff
@@ -115,15 +115,29 @@ export const divider = (): DividerBlock => {
...
+                "text": `You've completed ${completedCount} action items`
+            }
+        },
+        {
+            "type": "section",
+            "text": {
+                "type": "mrkdwn",
+                "text": `On average, you take this long to complete action items: ${avg}`
+            }
+        },
```
Comment (tchupp) [inline src/view.ts:133]: Should this section show up if we have completed 0 items?

File: migrations/1593194295169_add-status-to-action-items.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable @typescript-eslint/camelcase */
+import { MigrationBuilder, ColumnDefinitions } from 'node-pg-migrate';
+
+export const shorthands: ColumnDefinitions | undefined = undefined;
+
+export async function up(pgm: MigrationBuilder): Promise<void> {
+    pgm.addColumn('action_items', {
+        status: { type: 'text' }
+    });
+}
```
Comment (tchupp) [inline migrations/1593194295169_add-status-to-action-items.ts:10]: Should this be nullable?
How will we handle the existing action items?

If we query with this `SELECT * FROM action_items WHERE workspace_id = $1 AND channel_id = $2 AND status='OPEN'`, we will effectively ignore the existing action items

Comment (chance-cyphers) [reply migrations/1593194295169_add-status-to-action-items.ts:10]: just a simple `update action_items set status="OPEN"` should do the trick

File: src/view.ts (hunk 4 above)
Comment (chance-cyphers) [reply src/view.ts:133]: probably not, but given our audience I'd almost rather they know some feature is there. Next thing I plan on working on is re-arranging the home page and making it not look like garbage

File: migrations/1593194295169_add-status-to-action-items.ts (hunk 5 above)
Comment (tchupp) [reply migrations/1593194295169_add-status-to-action-items.ts:10]: Could we make the column non-nullable then? And set the default value to `"OPEN"`?
IIRC, that will _atomically_ add the column and set the value

File: src/view.ts (hunk 4 above)
Comment (tchupp) [reply src/view.ts:133]: 👍🏼

**Pull request merged**